### PR TITLE
Added If Null operator (??)

### PR DESCRIPTION
**Pull request by mastodon0**
If null/coalesce operator for issue #11

### CODE REVIEW

File: lib/src/compiler/expression_parser/ast.dart [hunk 1]
```diff
@@ -74,6 +74,17 @@ class Conditional extends AST {
   }
 }
 
+class IfNull extends AST {
+  AST condition;
+  AST nullExp;
+  IfNull(this.condition, this.nullExp) : super() {
+    /* super call moved to initializer */;
```
Comment (matanlurey) [inline lib/src/compiler/expression_parser/ast.dart]: This was ugly code for TS, you can remove it, just

```dart
IfNull(this.condition, this.nullExp);
```

File: lib/src/compiler/expression_parser/ast.dart [hunk 2]
```diff
@@ -74,6 +74,17 @@ class Conditional extends AST {
   }
 }
 
+class IfNull extends AST {
+  AST condition;
```
Comment (matanlurey) [inline lib/src/compiler/expression_parser/ast.dart]: Let's properly make both of these fields final.

File: lib/src/compiler/output/output_ast.dart [hunk 3]
```diff
@@ -361,6 +365,18 @@ class ConditionalExpr extends Expression {
   }
 }
 
+class IfNullExpr extends Expression {
```
Comment (matanlurey) [inline lib/src/compiler/output/output_ast.dart]: Ditto above, just a little documentation.

File: lib/src/compiler/output/output_ast.dart [hunk 4]
```diff
@@ -361,6 +365,18 @@ class ConditionalExpr extends Expression {
...
+  Expression condition;
```
Comment (matanlurey) [inline lib/src/compiler/output/output_ast.dart]: final both of these fields

File: lib/src/compiler/output/output_ast.dart [hunk 5]
```diff
@@ -361,6 +365,18 @@ class ConditionalExpr extends Expression {
...
+  Expression nullCase;
+  IfNullExpr(this.condition, Expression nullCase, [OutputType type = null])
+      : super(isPresent(type) ? type : nullCase.type) {
+    /* super call moved to initializer */;
```
Comment (matanlurey) [inline lib/src/compiler/output/output_ast.dart]: We probably don't need this either.

File: lib/src/compiler/expression_parser/ast.dart [hunk 6]
```diff
@@ -74,6 +74,17 @@ class Conditional extends AST {
   }
 }
 
+class IfNull extends AST {
```
Comment (matanlurey) [inline lib/src/compiler/expression_parser/ast.dart]: Mind leaving a tiny bit of dartdoc here:

/// Represents the `??` expression in Dart

**Comment by matanlurey** [comment]
Can you double check this works with the transformer? 

(Unfortunately our tests aren't very good right now)

**Comment by mastodon0** [comment]
Thank you for the advice, I hope this is more like you would have done it.
I checked the JS output, it looked like `t1 == null ? "unknown value" : t1`.

**Comment by kevmoo** [comment]
Please rebase on the latest `master` – we now have travis enabled.

File: lib/src/compiler/output/abstract_emitter.dart [hunk 7]
```diff
@@ -321,6 +321,16 @@ abstract class AbstractEmitterVisitor
     return null;
   }
 
+  dynamic visitIfNullExpr(o.IfNullExpr ast, dynamic context) {
+    EmitterVisitorContext ctx = context;
+    ctx.print('''(''');
```
Comment (chalin) [inline lib/src/compiler/output/abstract_emitter.dart]: Using triple quotes here serves no purpose. Single quotes will do, and can be used here (Line 326), as well as lines 328 and 330.

**Comment by kevmoo** [comment]
@mastodon0 care to do another rebase? This should fix the tests

**Pull request closed without merging**